packages/block-editor/src/components/global-styles/hooks.js

```diff
@@ -99,47 +188,151 @@ export function useGlobalStyle( path, blockName, source = 'all' ) {
 			set(
 				newUserConfig,
 				finalPath,
-				getPresetVariableFromValue(
-					mergedConfig.settings,
-					blockName,
-					path,
-					newValue
-				)
+				shouldDecodeEncode
+					? getPresetVariableFromValue(
+							mergedConfig.settings,
+							blockName,
+							path,
+							newValue
+					  )
+					: newValue
 			);
 			return newUserConfig;
 		} );
 	};
 
-	let result;
+	let rawResult, result;
 	switch ( source ) {
 		case 'all':
-			result = getValueFromVariable(
-				mergedConfig,
-				blockName,
+			rawResult =
 				// The stlyes.css path is allowed to be empty, so don't revert to base if undefined.
 				finalPath === 'styles.css'
 					? get( userConfig, finalPath )
 					: get( userConfig, finalPath ) ??
-							get( baseConfig, finalPath )
-			);
+					  get( baseConfig, finalPath );
+			result = shouldDecodeEncode
+				? getValueFromVariable( mergedConfig, blockName, rawResult )
+				: rawResult;
 			break;
 		case 'user':
-			result = getValueFromVariable(
-				mergedConfig,
-				blockName,
-				get( userConfig, finalPath )
-			);
+			rawResult = get( userConfig, finalPath );
+			result = shouldDecodeEncode
+				? getValueFromVariable( mergedConfig, blockName, rawResult )
+				: rawResult;
 			break;
 		case 'base':
-			result = getValueFromVariable(
-				baseConfig,
-				blockName,
-				get( baseConfig, finalPath )
-			);
+			rawResult = get( baseConfig, finalPath );
+			result = shouldDecodeEncode
+				? getValueFromVariable( baseConfig, blockName, rawResult )
+				: rawResult;
 			break;
 		default:
 			throw 'Unsupported source';
 	}
 
 	return [ result, setStyle ];
```

Comment:
It looks like the logic shouldDecodeEncode ? getValueFromVariable( ) : rawResult is repeated in each case of the switch statement. To avoid this repetition, it would make sense to move this logic outside the switch statement and perform it only once after the switch.